src/Lean/Language/Lean.lean

```diff
@@ -0,0 +1,471 @@
+input `def a := b private def c`. If we remove the space after `private`, the two commands
+syntactically become one with an application of `privatedef` to `b` even though the edit was
+strictly after the end of the first command.
+
+So obviously we must include at least the extent of the token that made the parser stop parsing a
+command as well such that invalidating the private token invalidates the preceding command.
+Unfortunately this is not sufficient either, given the following input:
+```
+structure a where /-- b -/ @[c] private axiom d : Nat
+```
+This is a syntactically valid sequence of a field-less structure and a declaration. If we again
+delete the space after private, it becomes a syntactically correct structure with a single field
+privateaxiom! So clearly, because of uses of atomic in the grammar, an edit can affect a command
+syntax tree even across multiple tokens.
+
+Now, what we do today, and have done since Lean 3, is to always reparse the last command completely
+preceding the edit location. If its syntax tree is unchanged, we preserve its data and reprocess all
+following commands only, otherwise we reprocess it fully as well. This seems to have worked well so
+far but it does seem a bit arbitrary given that even if it works for our current grammar, it can
+certainly be extended in ways that break the assumption.
+
+Finally, a more actually principled and generic solution would be to invalidate a syntax tree when
+the parser has reached the edit location during parsing. If it did not, surely the edit cannot have
+an effect on the syntax tree in question. Sadly such a "high-water mark" parser position does not
+exist currently and likely it could at best be approximated by e.g. "furthest `tokenFn` parse". Thus
+we remain at "go two commands up" at this point.
+-/
+
+set_option linter.missingDocs true
+
+namespace Lean.Language.Lean
+open Lean.Elab
+open Lean.Parser
+
+private def pushOpt (a? : Option α) (as : Array α) : Array α :=
+  match a? with
+  | some a => as.push a
+  | none   => as
+
+/-- Option for capturing output to stderr during elaboration. -/
+register_builtin_option stderrAsMessages : Bool := {
+  defValue := false
+  group    := "server"
+  descr    := "(server) capture output to the Lean stderr channel (such as from `dbg_trace`) during elaboration of a command as a diagnostic message"
+}
+
+/-- Option for showing elaboration errors from partial syntax errors. -/
+register_builtin_option showPartialSyntaxErrors : Bool := {
+  defValue := false
+  descr    := "show elaboration errors from partial syntax trees (i.e. after parser recovery)"
+}
+
+/-! The hierarchy of Lean snapshot types -/
+
+/-- Final state of processing of a command. -/
+structure CommandFinishedSnapshot extends Snapshot where
+  /-- Resulting elaboration state. -/
+  cmdState : Command.State
+deriving Nonempty
+instance : ToSnapshotTree CommandFinishedSnapshot where
```

Comment:
Intentional here as the instance is closely related to the structure's structure (so much so that it could become a derive handler) but I suppose that is subjective